File: 2019/day1/day1.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script
module Day1 where

import System.Environment (getArgs)

fuelReqs :: Int -> Int
fuelReqs x = x `quot` 3 - 2

realFuelReqs :: Int -> Int
realFuelReqs x
  | x < 0     = 0
  | otherwise = next + realFuelReqs next
  where next = x `quot` 3 - 2

star1 :: [Int] -> Int
star1 = sum . map fuelReqs

star2 :: [Int] -> Int
star2 = sum . map realFuelReqs

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  input <- map read . lines <$> readFile filename
  print $ star1 input
  print $ star2 input

File: 2019/day1/day1.rs
#!/usr/bin/env run-cargo-script
fn fuel_reqs(mass: i32) -> i32 {
    mass / 3 - 2
}

fn real_fuel_reqs(mass: i32) -> i32 {
    if mass < 0 {
        0
    } else {
        let fuel = mass / 3 - 2;
        fuel + real_fuel_reqs(fuel)
    }
}

fn star1(input: &Vec<i32>) -> i32 {
    input.iter().map(|i| fuel_reqs(*i)).sum()
}

fn star2(input: &Vec<i32>) -> i32 {
    input.iter().map(|i| real_fuel_reqs(*i)).sum()
}

fn main() {
    let args: Vec<_> = std::env::args().collect();
    let filename = args.get(1).map(|s| &s[..]).unwrap_or("input.txt");
    let input = std::fs::read_to_string(filename)
        .unwrap()
        .split_whitespace()
        .map(|i| i.parse().unwrap())
        .collect();

    println!("{}", star1(&input));
    println!("{}", star2(&input));
}
